### PR TITLE
Peepmatic: Do not use paths in linear IR

### PR DESCRIPTION
**Pull request by fitzgen**
Rather than using paths from the root instruction to the instruction we are matching against or checking if it is constant or whatever, use temporary variables. When we successfully match an instruction's opcode, we simultaneously define these temporaries for the instruction's operands. This is similar to how open-coding these matches in Rust would use `match` expressions with pattern matching to bind the operands to variables at the same time as switching on the opcode.

This shaves off about 1.8% of instructions retired when Peepmatic is enabled.

@abrown would you be comfortable reviewing this? If not, no worries, I can find another reviewer. Thanks!

### CODE REVIEW

File: cranelift/codegen/src/peepmatic.rs [hunk 1]
```diff
@@ -681,39 +651,290 @@ unsafe impl<'a, 'b> InstructionSet<'b> for &'a dyn TargetIsa {
...
+                imm,
+            } => {
+                operands.extend(iter::once(imm.into()));
+                opcode
+            }
+            ref otherwise => {
+                log::trace!("Not supported by Peepmatic: {:?}", otherwise);
+                return None;
+            }
+        })
```
Comment (abrown) [inline cranelift/codegen/src/peepmatic.rs:937]: I mean, I think I understand why this is necessary but it is a shame not to be able to use [opcode](https://docs.rs/cranelift-codegen/0.66.0/cranelift_codegen/ir/instructions/enum.InstructionData.html#method.opcode) and [arguments](https://docs.rs/cranelift-codegen/0.66.0/cranelift_codegen/ir/instructions/enum.InstructionData.html#method.arguments) off of `InstructionData` to make things a bit less verbose.

File: cranelift/peepmatic/crates/test/tests/tests.rs [hunk 2]
```diff
@@ -288,7 +288,7 @@ fn subsumption() {
         vec![expected_lhs, expected_rhs],
     );
 
-    let new = optimizer.apply_one(&mut program, iadd);
+    let new = dbg!(optimizer.apply_one(&mut program, iadd));
```
Comment (abrown) [inline cranelift/peepmatic/crates/test/tests/tests.rs:291]: Remove?

**Review by abrown — APPROVED** [review]
Looks good to me; I think both the `Match/Action` and `Lhs/Rhs` separations make things easier to understand.

File: cranelift/codegen/src/peepmatic.rs (hunk 1 above)
Comment (fitzgen) [reply cranelift/codegen/src/peepmatic.rs:937]: There are two reasons it is the way it currently is:

1. As written, we avoid matching on the instruction data twice. I've found that sometimes LLVM can collapse these back-to-back matches into a single match, but not always and it's kind of fragile.

2. The branching instructions supported by peepmatic ignore the block arguments, so we can't use `arguments` directly, we would still need to special case branches.

File: cranelift/peepmatic/crates/test/tests/tests.rs (hunk 2 above)
Comment (fitzgen) [reply cranelift/peepmatic/crates/test/tests/tests.rs:291]: good eye! thanks

**Pull request merged**